flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -0,0 +1,256 @@
+				}
+			}
+			this.inputsFromState = null;
+		}
+
+		buffer.startEmitterThread();
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		AsyncCollector<IN, OUT> collector = buffer.add(element);
+		userFunction.asyncInvoke(element.getValue(), collector);
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		buffer.add(mark);
+	}
+
+	@Override
+	public void processLatencyMarker(LatencyMarker latencyMarker) throws Exception {
+		buffer.add(latencyMarker);
+	}
+
+	@Override
+	public void snapshotState(FSDataOutputStream out, long checkpointId, long timestamp) throws Exception {
+		List<StreamElement> elements = buffer.getStreamElementsInBuffer();
+
+		serializeStreamElements(elements, out);
+	}
+
+	@Override
+	public void restoreState(FSDataInputStream in) throws Exception {
+		this.inputsFromState = deserializeStreamElements(in);
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+
+		buffer.waitEmpty();
+		buffer.stopEmitterThread();
+	}
+
+	@Override
+	public void dispose() throws Exception {
+		super.dispose();
+
+		buffer.stopEmitterThread();
+	}
+
+	public void sendLatencyMarker(LatencyMarker marker) throws Exception {
+		super.processLatencyMarker(marker);
+	}
+
+	private void serializeStreamElements(List<StreamElement> input,
+										FSDataOutputStream stream) throws IOException {
+		stream.write(input.size());
+
+		for (StreamElement element : input) {
+			if (element.isRecord()) {
+				stream.write(1);
+
+				StreamRecord<IN> record = element.asRecord();
+
+				outputSerializer.clear();
+				inTypeSerializer.serialize(record.getValue(), outputSerializer);
+				outputSerializer.writeLong(record.getTimestamp());
+				outputSerializer.writeBoolean(record.hasTimestamp());
+
+				stream.write(outputSerializer.getCopyOfBuffer());
+			}
+			else if (element.isWatermark()) {
+				stream.write(0);
+
+				Watermark watermark = element.asWatermark();
+
+				outputSerializer.clear();
+				outputSerializer.writeLong(watermark.getTimestamp());
+
+				stream.write(outputSerializer.getCopyOfBuffer());
+			}
+			else if (element.isLatencyMarker()) {
+				stream.write(2);
+
+				LatencyMarker marker = element.asLatencyMarker();
+
+				outputSerializer.clear();
+				outputSerializer.writeLong(marker.getMarkedTime());
+				outputSerializer.writeInt(marker.getVertexID());
+				outputSerializer.writeInt(marker.getSubtaskIndex());
+
+				stream.write(outputSerializer.getCopyOfBuffer());
+			}
+			else {
+				throw new IOException("Unknown element type: "+element);
+			}
+		}
+	}
+
+	private List<StreamElement> deserializeStreamElements(FSDataInputStream stream) throws IOException {
+		DataInputViewStreamWrapper wrapper = new DataInputViewStreamWrapper(stream);
+
+		int size = wrapper.read();
+
+		List<StreamElement> ret = new ArrayList<>(size);
+		for (int i = 0; i < size; ++i) {
+			int flag = wrapper.read();
+
+			if (flag == 1) {
```

Comment:
Reply to the serialize and deserialize part. StreamElementSerializer is so cool :-)